json_rpc/router.nim

```diff
@@ -12,8 +12,11 @@ type
   # Procedure signature accepted as an RPC call by server
   RpcProc* = proc(input: JsonNode): Future[StringOfJson] {.gcsafe, raises: [Defect, CatchableError].}
 
+  ProxyCall* = proc(name: string, input: JsonNode): Future[JsonNode] {.gcsafe, raises: [Defect, CatchableError].}
+
   RpcRouter* = object
     procs*: Table[string, RpcProc]
+    proxyCalls*: Table[string, ProxyCall]
```

Comment:
Adding this table to the router gets the job done, but it's a bit awkward specialization for such a general purpose library. A more general design is possible if you provide the users with an API allowing them to handle "raw" JSON RPC messages (i.e. one where the input is the compete `JsonNode` of the request). This can either be a callback that sits before the router in the framework or you can provide some building blocks that will make it easy for the users to implement their own reader loop handling the JSON-RPC messages one by one.

If such provisions were available, implementing the proxy would be a matter of gluing together some of the lower-level components in the user application (or in a helper module such as `rpcproxy`).

Reply:
I kinda agree with you about this specialisation issue (or maybe it is just the question of naming, maybe it should be names `MissingMethodHandler` ).

My point is that I see router as this kinda glue layer, which enables user of the libary to connect server with other stuff, and `rpcproxy` as it currently stands just glues http server with client + some methods to make everything more easy to use, and I am not sure that adding another level of indirection is necessary here.

My another approach was to make router polymorphic i.e have some base class with some of the basic methods and one method to handle `JsonNode` of the request which would need to be overriden in subclasses. Then we could have basic router as it is now, we could have `ProxyRouter` which could have this addtional table, and user of library could implement its own router with whatever abilities he wants. But I ultimately deemed this approach to too complex for current needs. 

Either way let my know what you think about polymorphic approach and I will try to experiment with approach you suggested.

Reply:
Indeed, calling this a `MissingMethodHandler` probably hits the sweet spot of being general purpose and practical enough for inclusion in the library. But then, it would make more sense to route all missing methods to the handler. You wouldn't have these lookups by method name (for the proxy calls).

Reply:
Having it in the table solves some other problems:
- user of router needs deliberately decide which methods will use missing method handler and which will use normal rpc call, which disables possibility of proxying some unwanted methods by accident. (as all other methods will return missing method error)
- user can in theory have different handlers for different calls (so lets if some endpoints require addtional handling, formating , logging etc)

Reply:
For me, a `MissingMethodHandler` will be responsible for handling all requests that doesn't have other handlers installed. Some people will be familiar with the concept form Ruby's [method_missing](https://apidock.com/ruby/BasicObject/method_missing) feature which is consistent with such a behavior. 

Otherwise, what's the difference between a RPC handler and a proxy handler? Instead of having a separate proxy table, the regular RPC handlers can accept a `JsonNode` argument which would skip the automated parameter marshaling logic. This, combined with the ability to make "raw" requests by passing in a `JsonNode` in `client.call(name, params)`, would be another way to implement a proxy.

Reply:
you are totally right ! My confusion was from the fact that I missed when the serialization/deserialization was really taking place, but as you mention current rpc proc (not macro) deals with raw request, so it is entirely possible to implement proxy by using current router capabilities by just having partially applied proc which returns correct `RpcProc` type. Thanks for discussion :+1:

so in summary:
- I have removed the additional table as it is not necessary
- I have left proxy implementation as it nice place to glue all the code
- I have removed missing handler as it is not neccessary to proxy implementation, if you think it would be valuable addition I can add it in spearate pr.